File: desktop/src/main/java/dev/ultreon/quantum/desktop/DesktopGLWindow.java
package dev.ultreon.quantum.desktop;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window;
import org.lwjgl.glfw.GLFW;
import org.lwjgl.glfw.GLFWNativeCocoa;
import org.lwjgl.glfw.GLFWNativeWin32;

public class DesktopGLWindow extends DesktopWindow {
    private final Lwjgl3Window window;
    private String title;
    private final int[] xPos = new int[1];
    private final int[] yPos = new int[1];

    public DesktopGLWindow(Lwjgl3Window window) {
        this.window = window;
    }

    @Override
    public void update() {
        super.update();

        if (isDragging()) {
            GLFW.glfwGetWindowPos(getHandle(), xPos, yPos);
        }
    }

    @Override
    public long getHandle() {
        return window.getWindowHandle();
    }

    @Override
    public boolean isHovered() {
        return true;
    }

    @Override
    public boolean isMinimized() {
        return GLFW.glfwGetWindowAttrib(getHandle(), GLFW.GLFW_ICONIFIED) == GLFW.GLFW_TRUE;
    }

    @Override
    public boolean isMaximized() {
        return !System.getProperty("os.name").equals("Mac OS X") && GLFW.glfwGetWindowAttrib(getHandle(), GLFW.GLFW_MAXIMIZED) == GLFW.GLFW_TRUE;
    }

    @Override
    public boolean isFocused() {
        return GLFW.glfwGetWindowAttrib(getHandle(), GLFW.GLFW_FOCUSED) == GLFW.GLFW_TRUE;
    }

    @Override
    public void close() {
        GLFW.glfwSetWindowShouldClose(getHandle(), true);
    }

    @Override
    public void setVisible(boolean visible) {
        if (visible) {
            GLFW.glfwShowWindow(getHandle());
        } else {
            GLFW.glfwHideWindow(getHandle());
        }
    }

    @Override
    public void minimize() {
        GLFW.glfwIconifyWindow(getHandle());
    }

    @Override
    public void maximize() {
        GLFW.glfwMaximizeWindow(getHandle());
    }

    @Override
    public void restore() {
        if (!System.getProperty("os.name").equals("Mac OS X")) {
            GLFW.glfwSetWindowAttrib(getHandle(), GLFW.GLFW_MAXIMIZED, GLFW.GLFW_FALSE);
            GLFW.glfwSetWindowAttrib(getHandle(), GLFW.GLFW_ICONIFIED, GLFW.GLFW_FALSE);
        }

        GLFW.glfwRestoreWindow(getHandle());
    }

    @Override
    public void focus() {
        GLFW.glfwFocusWindow(getHandle());
    }

    @Override
    public void requestAttention() {
        GLFW.glfwRequestWindowAttention(getHandle());
    }

    @Override
    public void setResizable(boolean resizable) {
        GLFW.glfwSetWindowAttrib(getHandle(), GLFW.GLFW_RESIZABLE, resizable ? GLFW.GLFW_TRUE : GLFW.GLFW_FALSE);
    }

    @Override
    public long getPeer() {
        if (System.getProperty("os.name").startsWith("Windows")) {
            return GLFWNativeWin32.glfwGetWin32Window(this.getHandle());
        } else if (System.getProperty("os.name").startsWith("Mac")) {
            return GLFWNativeCocoa.glfwGetCocoaWindow(this.getHandle());
        } else {
            return -1L;
        }
    }

    @Override
    public void setTitle(String title) {
        Gdx.graphics.setTitle(title);
        this.title = title;
    }

    @Override
    public String getTitle() {
        return title;
    }

    @Override
    public void setDragging(boolean dragging) {
        // no-op
    }
}

File: desktop/src/main/java/dev/ultreon/quantum/desktop/DesktopGLWindow.kt
package dev.ultreon.quantum.desktop

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window
import dev.ultreon.mixinprovider.PlatformOS
import org.lwjgl.glfw.GLFW
import org.lwjgl.glfw.GLFWNativeCocoa
import org.lwjgl.glfw.GLFWNativeWin32

class DesktopGLWindow(private val window: Lwjgl3Window) : DesktopWindow() {
  private var title: String? = null
  private val xPos = IntArray(1)
  private val yPos = IntArray(1)

  override fun update() {
    super.update()

    if (isDragging) {
      GLFW.glfwGetWindowPos(handle, xPos, yPos)
    }
  }

  override fun getHandle(): Long = window.windowHandle
  override fun isHovered(): Boolean = true
  override fun isMinimized(): Boolean = GLFW.glfwGetWindowAttrib(handle, GLFW.GLFW_ICONIFIED) == GLFW.GLFW_TRUE
  override fun isMaximized(): Boolean =
    !PlatformOS.isMac && GLFW.glfwGetWindowAttrib(handle, GLFW.GLFW_MAXIMIZED) == GLFW.GLFW_TRUE

  override fun isFocused(): Boolean = GLFW.glfwGetWindowAttrib(handle, GLFW.GLFW_FOCUSED) == GLFW.GLFW_TRUE
  override fun close() = GLFW.glfwSetWindowShouldClose(handle, true)
  override fun setVisible(visible: Boolean) = if (visible) GLFW.glfwShowWindow(handle) else GLFW.glfwHideWindow(handle)
  override fun minimize() = GLFW.glfwIconifyWindow(handle)
  override fun maximize() = GLFW.glfwMaximizeWindow(handle)

  override fun restore() {
    if (!PlatformOS.isMac) {
      GLFW.glfwSetWindowAttrib(handle, GLFW.GLFW_MAXIMIZED, GLFW.GLFW_FALSE)
      GLFW.glfwSetWindowAttrib(handle, GLFW.GLFW_ICONIFIED, GLFW.GLFW_FALSE)
    }

    GLFW.glfwRestoreWindow(handle)
  }

  override fun focus() = GLFW.glfwFocusWindow(handle)
  override fun requestAttention() = GLFW.glfwRequestWindowAttention(handle)
  override fun setResizable(resizable: Boolean) =
    GLFW.glfwSetWindowAttrib(handle, GLFW.GLFW_RESIZABLE, if (resizable) GLFW.GLFW_TRUE else GLFW.GLFW_FALSE)

  override fun getPeer(): Long =
    if (PlatformOS.isWindows) GLFWNativeWin32.glfwGetWin32Window(this.handle)
    else if (PlatformOS.isMac) GLFWNativeCocoa.glfwGetCocoaWindow(this.handle)
    else -1L

  override fun setTitle(title: String) {
    Gdx.graphics.setTitle(title)
    this.title = title
  }

  override fun getTitle(): String = title!!

  override fun setDragging(dragging: Boolean) {

  }
}
